### DISCUSSION TITLE
Upload workload to tenant

### DISCUSSION THREAD

**Discussion by dhrumil83** [microsoft/Microsoft-Fabric-workload-development-sample#15] [Q&A] [answered]
Uploading the nupgk gives this error, and there seems to be no messages in the audit logs that can point to what's missing.
https://learn.microsoft.com/en-us/fabric/workload-development-kit/manage-workload

![image (1)](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/14285436/9fa4389f-24c1-4cdc-a9f4-ee4b8f7681f5)

Followed these steps to upload FE and BE: https://learn.microsoft.com/en-us/fabric/workload-development-kit/azure-web-app-deployment-tutorial
Then tried to upload the nupkg file, so the workload/app is available across our tenant, but that does not seem to work.

Might have missed something, but hard to tell without info in the logs.

My account is missing the workloads tab as shown in the docs. Not clear how to enable that, and maybe that is causing this to fail?
![Screenshot 2024-06-10 at 10 31 14 AM](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/14285436/c9ff5b04-7286-4f3a-8caf-b000fc2e1097)

In dev mode, I do see the workload section, but not workload tab.
![Screenshot 2024-06-10 at 10 37 26 AM](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/14285436/8dee8bed-4411-487e-a76f-22328907904f)

**Comment by anitamayorov** [comment]
Hi @dhrumil83,
Please make sure your package is configured correctly - 
1. NuGet package structure - https://learn.microsoft.com/en-us/fabric/workload-development-kit/workload-environment#workload-nuget-package-structure
2. XMLs - must conform to the corresponding XSDs, 
https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/blob/main/Backend/src/Packages/manifest/WorkloadDefinition.xsd
https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/blob/main/Backend/src/Packages/manifest/ItemDefinition.xsd

3. You can verify package structure using the local development scenario, just note the differences of AADApp configurations (e.g., ResourceId)

Regarding WorkloadHub - it is the expected experience for now since no public Workloads available yet (even if there are activated Workloads in the tenant). We are consulting with our Product team on whether this experience matches the developers' requirement, will update.

**Reply by dhrumil83** [reply]
I am using the files generated via the build step in the sample, and then made changes to workloadmanifest.xml and manifest json file based on FE, BE and AADApp.
Item1.xml does not have much, since I have not added anything to BE yet, and outside of config changes, kept it as is.

Local:
![Screenshot 2024-06-17 at 4 57 31 PM](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/14285436/30c7e2f5-9c1b-4341-bdad-bc9bdc162fe9)

Cloud:
![Screenshot 2024-06-17 at 5 01 21 PM](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/14285436/0e6c5616-54ba-4966-ae62-d114ca41095d)

Both these packages fail to upload. Is there a place where I can upload these nupkg files for you to check out?
Also, I am on a trial account, and not sure if that matters?

**Reply by anitamayorov** [reply]
Hi @dhrumil83,

Are you trying to upload via Admin Portal? If so, your AADApp configuration must match your deployed BE and FE WebApps. 
Also, I see in second screenshot you are missing the FE/Item1.json file - this is needed for the FE resources of that item (parallel to Item1.xml in BE resources).

To help investigate this further, can you please:
1. Attach a requestId for the failed 'Upload' call, you can find it in DevTools/Network tab. 
2. Attach your WorkloadManifest.xml.

**Reply by dhrumil83** [reply]
Yes, trying to upload via Admin portal.
I did upload FE and BE to Azure, and put the links in the WorkloadManifest.xml file. Added item1.json for the release mode, but it did not help. Same error on upload.

Request id: 53123b33-52a0-d518-f354-8cd0e36cfeb1
![image (1)](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/14285436/903a74f5-a7d4-4847-9112-4bb076169a6c)
![image](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/14285436/0df77834-1dea-46dc-a271-1307eb48d7e5)



{"error":{"code":"ExtensionsInvalidPackage","pbi.error":{"code":"ExtensionsInvalidPackage","details":[{"code":"Reason","detail":{"type":1,"value":"InvalidParameter"}},{"code":"Target","detail":{"type":1,"value":"ResourceId"}}],"exceptionCulprit":1}}}

For resourceId, not clear from the docs if this value should be something else?
`api://localdevinstance`

`<ResourceId>api://localdevinstance/637a2d98-ac73-4698-aa2d-24ec7d560f05/Org.WorkloadSample</ResourceId>`

XML: 

```
<?xml version="1.0" encoding="utf-8" ?>
<WorkloadManifestConfiguration xmlns:xsi= "http://www.w3.org/2001/XMLSchema-instance" SchemaVersion="1.0.0">
  <Workload WorkloadName="Org.WorkloadSample" HostingType="Remote">
    <Version>1.0.0</Version>
    <RemoteServiceConfiguration>
      <CloudServiceConfiguration>
        <Cloud>Public</Cloud>
        <AADApp>
          <AppId>0a6c6188-6e81-4a76-83d0-7e58c9ad0f29</AppId> <!-- Your application ID -->
          <RedirectUri>https://esri-fmv-dev-fe.azurewebsites.net/close</RedirectUri>
          <ResourceId>api://localdevinstance/637a2d98-ac73-4698-aa2d-24ec7d560f05/Org.WorkloadSample</ResourceId>
        </AADApp>
        <Endpoints>
          <ServiceEndpoint>
            <Name>Workload</Name>
            <Url>https://esri-fmv-dev-be.azurewebsites.net</Url>
            <IsEndpointResolutionService>true</IsEndpointResolutionService>
            <EndpointResolutionContext>
              <ResolutionContextProperty>WorkspaceRegion</ResolutionContextProperty>
            </EndpointResolutionContext>
          </ServiceEndpoint>
          <ServiceEndpoint>
            <Name>Frontend</Name>
            <Url>https://esri-fmv-dev-fe.azurewebsites.net</Url>
            <IsEndpointResolutionService>false</IsEndpointResolutionService>
          </ServiceEndpoint>
        </Endpoints>
      </CloudServiceConfiguration>
    </RemoteServiceConfiguration>
  </Workload>
</WorkloadManifestConfiguration>

```

**Reply by anitamayorov** [reply]
Hi @dhrumil83,

Yes, the resourceId "api://localdevinstance/637a2d98-ac73-4698-aa2d-24ec7d560f05/Org.WorkloadSample" is only applicable to developer scenario, not the upload via Admin Portal scenario - therefore it fails. 
You should manage the resourceId in your Entra Id 'Expose an API' section 
![image](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/assets/80410523/efd204c6-b768-4435-83fc-2e5903fdf9e6)
https://learn.microsoft.com/en-us/fabric/workload-development-kit/authentication-tutorial#configure-your-application-in-microsoft-entra-id-manually

Also note, that when stating 

```          
<ServiceEndpoint>
            <Name>Workload</Name>
            <Url>https://esri-fmv-dev-be.azurewebsites.net</Url>
            <IsEndpointResolutionService>true</IsEndpointResolutionService>
            <EndpointResolutionContext>
              <ResolutionContextProperty>WorkspaceRegion</ResolutionContextProperty>
            </EndpointResolutionContext>
          </ServiceEndpoint>
```
You need to configure your endpoint resolution logic - otherwise you can state it as false. You can search for 'EndpointResolutionRequest' in our public Sample for more info.

**Reply by dhrumil83** [reply]
Changed IsEndpointResolutionService to false.

Changed ResourceId from `api://localdevinstance/637a2d98-ac73-4698-aa2d-24ec7d560f05/Org.WorkloadSample` to `api://esri-fmv-dev-be.azurewebsites.net/637a2d98-ac73-4698-aa2d-24ec7d560f05/Org.WorkloadSample`, but still the same error.
Also changed this under "Expose an API" section from localdevinstance to the BE url.
Requestid: 81e8cb7a-2e10-211c-3c63-b12318c89b3f

Also tried to remove ResourceId from the manifest file, but that did not help either. Different message, but I am assuming ResourceId is needed:
`{"error":{"code":"ExtensionsInvalidPackage","pbi.error":{"code":"ExtensionsInvalidPackage","details":[{"code":"Reason","detail":{"type":1,"value":"InvalidParameter"}},{"code":"Target","detail":{"type":1,"value":"WorklodDefinition"}}],"exceptionCulprit":1}}}`
Requestid: c84da8f9-c295-073f-bd53-0af8adc3dd7a

**Reply by anitamayorov** [reply]
Hi @dhrumil83, 

Your resourceId should look like this: 

```https://<ISV's tenant verified domain>/<workload frontend server>/<workload backend server>/<workload id>/<optional string>```

That's the first requestId - your resourceId as attached in your message has no https scheme, and no sever info 

Regarding the second requestId, resourceId under AADApp configuration in manifest is a mandatory field - you cannot omit it from the package.

**Reply by dhrumil83** [reply]
Thanks @anitamayorov 

Possible to do this for tenant level testing without setting up custom domain(using azurewebsites.net, where I have uploaded FE and BE)? 
Else can look into the domain setup.

**Reply by anitamayorov** [reply, marked as answer]
Hi @dhrumil83,
ResourceId property has only two 'modes': developer mode (with the 'localinstance') used for connecting a developer workload to Fabric, and production mode as described in the comment above. 
In order to upload and activate a workload in your tenant, you will need the production mode for ResourceId and rest of configurations in your XML files to be compatible with your Entra ID (AAD) configurations, including custom domains.

**Reply by dhrumil83** [reply]
Hi @anitamayorov

Got the domain setup, but still getting same error with ResourceId not valid
https://fabricdev.geocloud.com/fe/be/Org.WorkloadSample/preview

```
{
    "error": {
        "code": "ExtensionsInvalidPackage",
        "pbi.error": {
            "code": "ExtensionsInvalidPackage",
            "details": [
                {
                    "code": "Reason",
                    "detail": {
                        "type": 1,
                        "value": "InvalidParameter"
                    }
                },
                {
                    "code": "Target",
                    "detail": {
                        "type": 1,
                        "value": "ResourceId"
                    }
                }
            ],
            "exceptionCulprit": 1
        }
    }
}
```

Requestid:
8bab442c-19a7-5983-f19b-0cdb86f37a93

Attached is the nupkg file(zipped) that I am trying to upload.
[ManifestPackageRelease.1.0.0.nupkg.zip](https://github.com/user-attachments/files/16146446/ManifestPackageRelease.1.0.0.nupkg.zip)

**Reply by anitamayorov** [reply]
Hi @dhrumil83,

It looks like your domain is not part of your tenant's verified domains, you will need to configure it in Azure App Service Domain section. 
Can you please verify it is configured correctly and try and upload again?

**Reply by dhrumil83** [reply]
Thanks. That was it. 
Upload was successful 👍

**Reply by gregahren** [reply]
> Hi @dhrumil83, ResourceId property has only two 'modes': developer mode (with the 'localinstance') used for connecting a developer workload to Fabric, and production mode as described in the comment above. In order to upload and activate a workload in your tenant, you will need the production mode for ResourceId and rest of configurations in your XML files to be compatible with your Entra ID (AAD) configurations, including custom domains.

Is it possible to publish an unverified package to a tenant separate from the publishers tenant?

For example we have the publisher tenant be gtm.something.com, and we want to publish it to company.us tenant. 
From the documentation this limitation is not clear.

**Reply by anitamayorov** [reply]
Hi @gregahren, 
You can publish an unverified package to other tenants, see this documentation: https://learn.microsoft.com/en-us/fabric/workload-development-kit/more-workloads-add#adding-unverified-workloads. The other tenant admin should enable a tenant switch to allow this: "Users can see and work with uncertified partner workload"

**Reply by gregahren** [reply]
> Hi @gregahren, You can publish an unverified package to other tenants, see this documentation: https://learn.microsoft.com/en-us/fabric/workload-development-kit/more-workloads-add#adding-unverified-workloads. The other tenant admin should enable a tenant switch to allow this: "Users can see and work with uncertified partner workload"

To be able to see the workload, do we need a workload to be in public preview phase so it can be added to up to 10 tenants?

**Reply by anitamayorov** [reply]
Hi @gregahren 
The be able to publish to 10 consumer tenants you must go through the publishing process, you can see details here https://learn.microsoft.com/en-us/fabric/workload-development-kit/publish-workload-flow. 
If you activate the workload in your tenant, then it will be visible to your tenant only and you will be able to see the workload in the workloads hub, tagged with 'added by my organization', no publishing process is needed for this case.

**Reply by gregahren** [reply]
> Hi @gregahren The be able to publish to 10 consumer tenants you must go through the publishing process, you can see details here https://learn.microsoft.com/en-us/fabric/workload-development-kit/publish-workload-flow. If you activate the workload in your tenant, then it will be visible to your tenant only and you will be able to see the workload in the workloads hub, tagged with 'added by my organization', no publishing process is needed for this case.

But if we want to add it to another tenant we get this error when uploading workload:
{
    "error": {
        "code": "ExtensionsInvalidPackage",
        "pbi.error": {
            "code": "ExtensionsInvalidPackage",
            "details": [
                {
                    "code": "Reason",
                    "detail": {
                        "type": 1,
                        "value": "InvalidParameter"
                    }
                },
                {
                    "code": "Target",
                    "detail": {
                        "type": 1,
                        "value": "ResourceId"
                    }
                },
                **{
                    "code": "Message",
                    "detail": {
                        "type": 1,
                        "value": "ResourceId domain - other-tenants.com is not in the tenant domains list"
                    }
                }**
            ],
            "exceptionCulprit": 1
        }
    }
}

So lets say other-tenants.com domain is our tenant domain, and other tenant cant add it to tenant domain list, what can we do in that case?

**Reply by anitamayorov** [reply]
Hi @gregahren , 

1. ResourceId property should be configured in your Entra ID application and match **your** tenant's verified domain, see here https://learn.microsoft.com/en-us/fabric/workload-development-kit/workload-cloud-setup#microsoft-entra-id-app-resourceid-format.
2. The consumer tenant, say other-tenant.com, should not be configured in any of your manifests. You can publish to selected tenants in Admin portal when you open the side pane with all of your uploaded versions and details on which version is currently activated to tenant. For this option to be available, you need to go through our publishing process.
